source/Week_11_correction/Graph avec des classes - solution.ipynb

```diff
@@ -0,0 +1,180 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Exercice 1\n",
+    "\n",
+    "Pour conceptualiser notre graphes nous allons utliser deux classes diffenrentes, les sommets seront des chaines de caratères, indivuelement `vertex`;\n",
+    "+ une classe `Edge` pour definir les arretes.\n",
+    "Chaque instance aura comme attribut un sommet de départ `from_vertex`, un sommet d'arrivée `to_vertex` et un poids `weight`. La classe aura comme unique methode `__repr__`pour retourner l'objet sous la forme d'un dictionaire.\n",
+    "+ une classe `Graph`pour definir un graphe.\n",
+    "Chaque instance aura comme attribut un set `edges` contenant les arêtes (qui seront des objets) et un set `vertices` contenant les sommets. astuce: pour initialiser un set vide utilisez `set()`\n",
+    "\n",
+    "\n",
+    "La classe `Graph` aura comme méthodes:\n",
+    "+ `add_vertex` pour créer un sommet\n",
+    "+ `edge_exist` pour vérifier si une arête existe, elle retourne `weight` si oui, `False` sinon. \n",
+    "+ `update_weight` pour mettre le poids à jour d'une connexion existante, et appeler la méthode `new_edge` sinon.\n",
+    "+ `new_edge` pour créer une instance de `Edge`et l'ajouter au set `edges` si la connexion n'existe pas déjà, si elle existe avec un autre poids mettre à jour le poids, si elle existe de façon identique alors retournez le dans la console avec `print(...)`. Astuce: Ici il peut être judicieux de créer une méthode privée `__generate_edge` et de l'appeler puisque avant de créer une connexion il faut s'assurer que les sommets existes et les créer sinon. \n",
+    "+ `del_edge` pour supprimer un `edge`\n",
+    "+ `__repr__` pour retourner l'objet sous forme de dictionnaire. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "\n",
+    "class Edge:\n",
+    "    # votre code ici\n",
+    "    def __init__(self, from_vertex, to_vertex, weight):\n",
+    "        self.from_vertex = from_vertex\n",
+    "        self.to_vertex = to_vertex\n",
+    "        self.weight = weight\n",
+    "\n",
+    "    def __repr__(self):\n",
+    "        return str({\"from_vertex\": self.from_vertex, \"to_vertex\": self.to_vertex, \"weight\": self.weight})\n",
+    "\n",
+    "    \n",
+    "class Graph:\n",
+    "    # votre code ici\n",
+    "    def __init__(self):\n",
+    "        self.edges = set()\n",
+    "        self.vertices = set()\n",
+    "\n",
+    "    def add_vertex(self, name):\n",
+    "        self.vertices.add(name)\n",
+    "\n",
+    "    def edge_exist(self, from_vertex, to_vertex):\n",
+    "        for edge in self.edges:\n",
+    "            if edge.from_vertex is from_vertex and edge.to_vertex is to_vertex:\n",
+    "                return edge.weight\n",
```

Comment:
What happens if `edge.weight == 0`?

Reply:
there is no contraint regarding the Weight value, it could be a str or even a boolean